[Feature] Add signing methods for Plaintext and Record types

Roee-87 opened the pull request:
<!--
    Thank you for submitting the PR! We appreciate you spending the time to work on these changes.

    Help us understand your motivation by explaining why you decided to make this change.

    Happy contributing!
-->

## Motivation

Signatures generated from the SDK cannot be verified in Aleo programs.  Aleo programs are expecting signatures that sign field representations of the data.  The SDK is currently signing bytes.
This PR introduces methods for the PrivateKey and Signature types that sign message data as field representations.

## Test Plan

<!--
    If you changed any code,
    please provide us with clear instructions on how you verified your changes work.
    Bonus points for screenshots and videos!
-->

Generate a signature using the SDK and verify it in a Leo program.

1.  Build the SDK locally from this branch.
2.  Run `yarn install` from the root of the SDK directory and again from `sdk/create-leo-app/template-node` subdirectory.
4.  Replace the code in `index.js` with the following snippet and then run `yarn dev` to generate a schnorr signature.
```javascript
import {initThreadPool, PrivateKey} from "@provablehq/sdk/testnet.js";
import assert from "assert";

await initThreadPool();

function test_sig() {
    const privateKey = PrivateKey.from_string("APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH");
    const message = "42u8";
    const sig = privateKey.sign_fields(message);
    console.log("Signature: ", sig.to_string());

    console.log(sig.verify_fields(privateKey.to_address(), message));
}

test_sig();
```
5.  Create a new leo project using `leo new PROJECT_NAME` and replace the Leo code in `main.leo` with this code:
```Leo
program versig.aleo {
    @noupgrade
    async constructor() {}

    transition verify_signature(message: u8, pass: signature) -> address {
        let check: bool = pass.verify(self.signer, message);
        assert_eq(check, true);
        return self.signer;
    }
}
```
6.  Run the following command to verify the signature that you generated in step 4.
```bash
leo run versig.aleo/verify_signature 42u8 ${SIGNATURE}
```
Expected output:
```➕Adding programs to the VM in the following order:
  - versig.aleo (edition: 1)

➡️  Output

 • aleo1rhgdu77hgyqd3xjj8ucu3jj9r2krwz6mnzyd80gncr5fxcwlh5rsvzp9px
 ```
7.  If you replace the message with any other value than what was signed in step 4, the signature verification should fail.
```bash
➕Adding programs to the VM in the following order:
  - versig.aleo (edition: 1)
Failed to evaluate instruction (assert.eq r2 true;): 'assert.eq' failed: 'false' is not equal to 'true' (should be equal)
```

8.  Signing messages with `PrivateKey.sign(message)` instead of `PrivateKey.sign_fields(message)` will pass the verification step in the JS code snippet but will fail to verify in the Leo program.

Roee-87 committed:
added methods for signing over field array

Roee-87 committed:
added sign_fields method to PrivateKey impl

Roee-87 committed:
removed field conversion for PrivateKey.sign_fields() method

Roee-87 committed:
fixed typos and build errors

Roee-87 committed:
ran cargo fmt

Roee-87 commented:
Note:  CI failing on this check:
```bash
  1 failing

  1) NodeConnection
       waitForTransactionConfirmation
         should throw for a malformed tx ID:
     AssertionError: expected 'Transaction at1dl9lze8wscct0dee8x9tjn…' to include 'Malformed transaction ID'
      at Context.<anonymous> (file:///home/circleci/project/sdk/tmp/mainnet/tests/network-client.test.js:286:44)
      ```

alexpitsikoulis commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +48,18 @@ impl Signature {
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
+    /// @returns {Signature} Signature of the message
+    pub fn sign_fields(private_key: &PrivateKey, message: &str) -> Result<Self, String> {
```
Is there a reason this returns a `Result` when `sign` just returns a `Signature`?

iamalwaysuncomfortable replied:
@alexpitsikoulis because this is signing an Aleo datatype and if the input is not a valid Aleo type, then this method will fail.

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,15 @@ impl PrivateKey {
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
+    /// @returns {Signature} Signature of the message
+    pub fn sign_fields(&self, message: &str) -> Result<Signature, String> {
```
What this method actually does is sign an Aleo type and will fail if it's NOT a valid Aleo type. 

The conversion into field happens under the hood and isn't really exposed to users, and generally functions presented in an API shouldn't describe the imperative process they're encapsulating, they should describe the top level action being done.

Here what we're accomplishing is signing a valid Aleo Value type (i.e. a Plaintext, Record, or Future) and so it should be named that. Sign_fields would indicate you're signing an array of fields. 

```suggestion
    pub fn sign_value(&self, message: &str) -> Result<Signature, String> {
```

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,15 @@ impl PrivateKey {
         Signature::sign(self, message)
     }
 
+    /// Convert a message to field representation and sign the fields with a private key
```
```suggestion
    /// Sign an instance of an valid Aleo data type or record.
```

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,15 @@ impl PrivateKey {
 
+    /// Convert a message to field representation and sign the fields with a private key
+    ///
+    /// @param {PrivateKey} private_key The private key to sign the message with
```
There's no private key input here.

```suggestion
```

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,15 @@ impl PrivateKey {
+    /// Convert a message to field representation and sign the fields with a private key
+    ///
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
```
```suggestion
    /// @param {String} message The string representation of the Aleo datatype or record to sign.
```

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,15 @@ impl PrivateKey {
+    ///
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
+    /// @returns {Signature} Signature of the message
```
```suggestion
    /// @returns {Signature} Signature of the message.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +48,18 @@ impl Signature {
+    /// Convert a message to field representation and sign the fields with a private key
+    ///
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
```
```suggestion
    /// @param {String} message The string representation of the Aleo datatype or record to sign.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +48,18 @@ impl Signature {
 
+    /// Convert a message to field representation and sign the fields with a private key
+    ///
+    /// @param {PrivateKey} private_key The private key to sign the message with
```
```suggestion
    /// @param {PrivateKey} private_key The private key to sign the message with.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +48,18 @@ impl Signature {
         Self(SignatureNative::sign_bytes(private_key, message, &mut StdRng::from_entropy()).unwrap())
     }
 
+    /// Convert a message to field representation and sign the fields with a private key
```
```suggestion
    /// Sign an instance of an valid Aleo data type or record.
```

Roee-87 committed:
added unit test for PrivateKey.sign_fields()

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +48,18 @@ impl Signature {
+    /// @param {PrivateKey} private_key The private key to sign the message with
+    /// @param {String} message The message to sign
+    /// @returns {Signature} Signature of the message
+    pub fn sign_fields(private_key: &PrivateKey, message: &str) -> Result<Self, String> {
```
Please change this to sign_value per the reasoning above.

```suggestion
    pub fn sign_value(private_key: &PrivateKey, message: &str) -> Result<Self, String> {
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +86,17 @@ impl Signature {
+    /// @param {Address} address The address to verify the signature with
+    /// @param {String} message The message to verify
+    /// @returns {boolean} True if the signature is valid, false otherwise
+    pub fn verify_fields(&self, address: &Address, message: &str) -> Result<bool, String> {
```
```suggestion
    pub fn verify_value(&self, address: &Address, message: &str) -> Result<bool, String> {
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +86,17 @@ impl Signature {
         self.0.verify_bytes(address, message)
     }
 
+    /// Verify a signature of a message in field representation with an address
```
```suggestion
    /// Verify a signature over an Aleo datatype or record by an address.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +86,17 @@ impl Signature {
 
+    /// Verify a signature of a message in field representation with an address
+    ///
+    /// @param {Address} address The address to verify the signature with
```
```suggestion
    /// @param {Address} address The address to verify the signature with.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +86,17 @@ impl Signature {
+    /// Verify a signature of a message in field representation with an address
+    ///
+    /// @param {Address} address The address to verify the signature with
+    /// @param {String} message The message to verify
```
```suggestion
    /// @param {String} message The message to verify, this message must be the string representation of a valid Aleo datatype or record.
```

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +86,17 @@ impl Signature {
+    ///
+    /// @param {Address} address The address to verify the signature with
+    /// @param {String} message The message to verify
+    /// @returns {boolean} True if the signature is valid, false otherwise
```
```suggestion
    /// @returns {boolean} True if the signature is valid, false otherwise.
```

iamalwaysuncomfortable reviewed (changes requested):
I'm requiring a few structural changes to this (mainly the naming and description of the function being added), but otherwise this is what we needed. Also, excellent Pr description, this is the level of descriptiveness in a PR summary that we should always strive for.

Also, please add several tests of valid Aleo datatypes (and invalid types) and if you can, add a small markdown file to our docs showing how to use this. Your PR description is actually a good start for this. Thanks for the work!

We'll be releasing Tuesday, so if you finish this by Monday, it'll ship with the release.

Roee-87 committed:
changed method name from sign_filds() to sign_value() and fixed docstrings

Roee-87 committed:
added tests

Roee-87 commented:
@iamalwaysuncomfortable I addressed the docstring revision requests and added a unit test for fail cases.
fixed docstring issues and modified method names:  [51ed6f9](https://github.com/ProvableHQ/sdk/pull/1099/commits/51ed6f93b1ca35e09e39e466c62c0870af03e22c) 

added extra unit tests:  [cd23c73](https://github.com/ProvableHQ/sdk/pull/1099/commits/cd23c73fdb8f71ccf3a224e08f26e0d5db34bfd8)

Roee-87 committed:
ran cargo fmt

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -261,4 +269,19 @@ mod tests {
         }
     }
+
+    #[wasm_bindgen_test]
+    pub fn test_signature_over_fields() {
+        for _ in 0..ITERATIONS {
+            // Sample a new private key and message.
+            let private_key = PrivateKey::new();
+            let rand_val: u64 = StdRng::from_entropy().gen();
+            let message = format!("{rand_val}field");
+
+            // Sign the message.
+            let signature = private_key.sign_value(&message).unwrap();
+            // Check the signature is valid.
+            assert!(signature.verify_value(&private_key.to_address(), &message).unwrap());
+        }
+    }
```
Can you add extra tests that test signatures over:

* Integer, Group, and Scalar datatypes
* Valid aleo records
* Aleo structs
* Aleo arrays

People will most often be signing `records`, `structs`, or `arrays` so I'd like to sanity check these work.

Roee-87 replied:
done [cea1431](https://github.com/ProvableHQ/sdk/pull/1099/commits/cea14315ab8bce8a02998ac6352292e518f7b9ab)

Roee-87 committed:
added integer and type tests

Roee-87 committed:
replaced 'type': 'module'

Roee-87 committed:
added unit tests for structs and arrays

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -196,6 +220,20 @@ mod tests {
+  _version: 0u8.public
+}";
+    const NESTED_STRUCT: &str = "{ player: aleo13nnjqa7h2u4mpl95guz97nhzkhlde750zsjnw59tkgdwc85lyurs295lxc, health: 100u8, inventory: { coins: 5u32, snacks: { candies: 5u64, vegetals: 6u64 } }, secret: 2group, cipher: 2scalar, is_alive: true }";
+    const INVALID_NESTED_STRUCT: &str = "{ player: aleo13nnjqa7h2u4mpl95guz97nhzkhlde750zsjnw59tkgdwc85lyurs295lxc, health: 100u8, inventory: { coins: 5u32, snacks: { candies: 5u64, vegetals: 6u64 } }, secret: 2group, cipher: 2scalar, is_alive: false }";
```
Could you rebase on main and move these into the `utilities/test` folder?

* `Structs` should go in `utilities/test/plaintext.rs`
* `Records` should go into `utilities/test/records.rs`

`ITERATIONS` can stay where it is.

iamalwaysuncomfortable reviewed (commented):
One final nit about where the test vectors live, otherwise this is good to go.

iamalwaysuncomfortable commented on wasm/src/account/private_key.rs:
```diff
@@ -100,6 +100,14 @@ impl PrivateKey {
+    ///
+    /// @param {String} message The string representation of the Aleo datatype or record to sign.
+    /// @returns {Signature} Signature of the message.
+    pub fn sign_value(&self, message: &str) -> Result<Signature, String> {
```
Also ensure to `js_name` any `snake_case` method into `camelCase`.

```suggestion
    #[wasm_bindgen(js_name = signValue)]
    pub fn sign_value(&self, message: &str) -> Result<Signature, String> {
```

Roee-87 replied:
[3ce2b9b](https://github.com/ProvableHQ/sdk/pull/1099/commits/3ce2b9b7586510bee61cb410be3f87377a45437d)

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -48,6 +49,18 @@ impl Signature {
+    /// @param {PrivateKey} private_key The private key used to sign the message.
+    /// @param {String} message The string representation of the Aleo datatype or record to sign.
+    /// @returns {Signature} Signature of the message.
+    pub fn sign_value(private_key: &PrivateKey, message: &str) -> Result<Self, String> {
```
```suggestion
    #[wasm_bidgen(js_value = signValue)]
    pub fn sign_value(private_key: &PrivateKey, message: &str) -> Result<Self, String> {
```

Roee-87 replied:
[3ce2b9b](https://github.com/ProvableHQ/sdk/pull/1099/commits/3ce2b9b7586510bee61cb410be3f87377a45437d)

iamalwaysuncomfortable commented on wasm/src/account/signature.rs:
```diff
@@ -74,6 +87,17 @@ impl Signature {
+    /// @param {Address} address The address used to verify the signature.
+    /// @param {String} message The message to verify, which must be the string representation of a valid Aleo datatype or record.
+    /// @returns {boolean} True if the signature is valid, false otherwise.
+    pub fn verify_value(&self, address: &Address, message: &str) -> Result<bool, String> {
```
```suggestion
    #[wasm_bindgen(js_name = verifyValue)]
    pub fn verify_value(&self, address: &Address, message: &str) -> Result<bool, String> {
```

Roee-87 replied:
[3ce2b9b](https://github.com/ProvableHQ/sdk/pull/1099/commits/3ce2b9b7586510bee61cb410be3f87377a45437d)

iamalwaysuncomfortable reviewed (commented):
Also don't forget to do `yarn build:sdk-docs` before you merge.

A good follow-on PR to this would be to make a JS method that can accept any `Plaintext`, `RecordPlaintext` `Field`, `Group`, `Scalar`, `Integer` type.

The way to do this is in the `sdk/models` folder, create a PlaintextObject type that unions the above types then, if any of those types are passed in, handle the conversion to string.

Roee-87 committed:
moved structs and record strings into utilities/tests dirs

Roee-87 committed:
fixed package.json

Roee-87 committed:
added WASM aliases

Roee-87 committed:
rebuilt docs

Roee-87 committed:
ran cargo fmt

iamalwaysuncomfortable commented on docs/api_reference/sdk-src_wasm.md:
```diff
@@ -1863,6 +1892,19 @@ __*return*__ | [PrivateKey](sdk-src_wasm.md) | **
 
 ## Methods
 
+### `sign_value(message) ► Signature`
```
```suggestion
### `signValue(message) ► Signature`
```

iamalwaysuncomfortable commented on docs/api_reference/sdk-src_wasm.md:
```diff
@@ -3537,6 +3593,20 @@ Cryptographic signature of a message signed by an Aleo account
 
 ## Methods
 
+### `sign_value(private_key, message) ► Signature`
```
```suggestion
### `signValue(private_key, message) ► Signature`
```

iamalwaysuncomfortable commented on docs/api_reference/sdk-src_wasm.md:
```diff
@@ -3611,6 +3681,20 @@ __*return*__ | [Plaintext](sdk-src_wasm.md) | **
 
 ---
 
+### `verify_value(address, message) ► boolean`
```
```suggestion
### `verifyValue(address, message) ► boolean`
```

iamalwaysuncomfortable committed:
Change function name in docs.

Signed-off-by: Mike Turner <mike@provable.com>
